### PR TITLE
groupreduction and subgroupreduction

### PR DESCRIPTION
**Pull request by brabreda**
I am unsure why my previous PR closed but here are the changes.
- I added docs
- I added tests

It was my first time writing tests, and they passed. How are these tested on GPUs, and do I need more tests?

### CODE REVIEW

**Comment by vchuravy** [comment]
Thank you! As you found the notion of a subgroup is not necessarily consistent across backends. So I am wondering if we need to expose that as feature or if backends could use them to implement reduction.

So as an example could we "just" add `rval= @reduce(op, val)` to perform a workgroup level reduction?

What is `rval` across all members of a workgroup. Is it the same or may it divergent and only the first workitem will receive a correct answer?

Do we need a broadcast/shuffle operation? 

I don't have the answers to those questions. 

I think adding `@reduce` is definitely something we want to do, but subgroup ops may need more thought.

**Comment by brabreda** [comment]
- `rval=@reduce(op, val)` is possible. In this case we only use `localmem` and `@synchronize` for the reduction. 

- at the moment only the first thread of the group returns an item, but I think it would be possible to change this.

- I Think providing a neutral value will have better performance, but I am not sure. I Will check this. 

I will fix this, meanwhile we can think about the wraps. If we will introduce it and in what way.

**Comment by vchuravy** [comment]
Yeah requiring a neutral value sounds good.

Could not each backend choose to use warp level ops, iff available? Instead of having the user make this decision.

**Comment by brabreda** [comment]
The problem with that approach is that warp intrinsics don’t work for every type. 

We then would need something like this:

`__reduce(op, val::union{type1, …},)` for wraps
`__reduce(op, val,)` for everything else 

In this case you don’t override the function. I am also not sure if this kind of type specialisation works where it takes the first function when it is one of the types. You probs have more expertise on this.

would it be prefered that every thread returns the right rval, I do think in that case I Will have to rethink the warp reduction bases groupreduce.

**Comment by vchuravy** [comment]
> would it be prefered that every thread returns the right rval? I do think in that case I Will have to rethink the warp reduction bases groupreduce.

No it just needs to be documented :) and we might want to add a `@broadcast`.


So it looks like CUDA.jl makes this decision based on the element type?
https://github.com/JuliaGPU/CUDA.jl/blob/d57e02018ccf00ceda7672f5f4d98e2ceef9106d/src/mapreduce.jl#L177-L181

File: src/KernelAbstractions.jl [hunk 1]
```diff
@@ -139,10 +139,31 @@ function unsafe_free! end
...
+"""
+    @subgroupsize()
+
+    returns the GPUs subgroupsize.
+"""
+macro subgroupsize()
+    quote
+        $__subgroupsize()
+    end
+end
```
Comment (vchuravy) [inline src/KernelAbstractions.jl:154]:
```suggestion
```

File: src/reduce.jl [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+    return val
+end
+
+@inline function __groupreduce(__ctx__, op, val, neutral, ::Type{T}, ::Val{false}) where {T}
+    idx_in_group = @index(Local)
+    groupsize = @groupsize()[1]
+    
+    localmem = @localmem(T, groupsize)
```
Comment (vchuravy) [inline src/reduce.jl:72]: I see if we can do a subgroupreduce the memory we need here is much reduced.

File: src/reduce.jl [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+"""
+
+@groupreduce(op, val, neutral, use_subgroups)
+
+Reduce values across a block
+- `op`: the operator of the reduction
+- `val`: value that each thread contibutes to the values that need to be reduced
+- `netral`: value of the operator, so that `op(netural, neutral) = neutral``
+- `use_subgroups`: make use of the subgroupreduction of the groupreduction
```
Comment (vchuravy) [inline src/reduce.jl:27]: I see the value of having a common implementation.

So I would define:

```
@reduce(op, val, neutral)
__reduce(op, val, neutral)
```

And then maybe:

```
__subgroupreduce & __subgroupsize (No @ version)
__can_subgroup_reduce(T) = false
```

And then we could define:

```
__reduce(__ctx___, op, val, neutral, ::Val{true})
__reduce(__ctx___, op, val, neutral, ::Val{false})
```

as you have here.

```
function __reduce(op, val::T, neutral::T) where T
     __reduce(op, val, neutral, Val(__can_subgroup_reduce(T)))
end
```

**Comment by brabreda** [comment]
Yeah so CUDA(and i think most GPU backends make the decision based on type). I was thinking if something like this would be possible:

For the implementation without warps:
`
@reduce(op,val,neutral)
__reduce(op,val,neutral)
`
And every back-end like CUDAKernels just implements something like this:
`
__reduce(op,val::T,neutral) where {T <: Union{Float32, Float64, etc...}}
`

I tried It today and I get a lot of dynamic invocation errors. But I am not sure what's at fault.

**Comment by vchuravy** [comment]
I recommed `using Cthulhu, using CUDA` and then using `CUDA.device_code_typed interactive=true`

**Comment by brabreda** [comment]
Only the group reduction without subgroups remains. It is ready to merge, once merged I will add support for Subgroups to the CUDAkernels, MetalKernels, and OneAPI.

File: src/reduce.jl [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    groupsize = @groupsize()[1]
+    
+    localmem = @localmem(T, groupsize)
+
+    @inbounds localmem[idx_in_group] = val
+
+    # perform the reduction
+    d = 1
+    while d < groupsize
+        @synchronize()
```
Comment (vchuravy) [inline src/reduce.jl:30]: https://github.com/JuliaGPU/KernelAbstractions.jl/issues/262

**Comment by brabreda** [comment]
The API is indeed neutral. But indeed the problem now is the @synchronize() in while loops.

File: src/reduce.jl (hunk 4 above)
Comment (maleadt) [reply src/reduce.jl:29]: Workaround?

**Pull request closed without merging**